crates/nu-protocol/src/value/test_derive.rs

```diff
@@ -0,0 +1,375 @@
+        Span::test_data(),
+    )
+    .unwrap();
+    assert_eq!(expected, actual);
+
+    let expected = UnnamedFieldsStruct::value();
+    let actual =
+        UnnamedFieldsStruct::<f64>::from_value(UnnamedFieldsStruct::value(), Span::test_data())
+            .unwrap()
+            .into_value_unknown();
+    assert_eq!(expected, actual);
+}
+
+#[test]
+fn unnamed_fields_struct_missing_value() {
+    let value = Value::test_list(vec![]);
+    let res: Result<UnnamedFieldsStruct<f64>, _> =
+        UnnamedFieldsStruct::from_value(value, Span::test_data());
+    assert!(res.is_err());
+}
+
+#[test]
+fn unnamed_fields_struct_incorrect_type() {
+    // Should work for every type that is not a record.
+    let value = Value::test_nothing();
+    let res: Result<UnnamedFieldsStruct<f64>, _> =
+        UnnamedFieldsStruct::from_value(value, Span::test_data());
+    assert!(res.is_err());
+}
+
+#[derive(IntoValue, FromValue, Debug, PartialEq)]
+struct UnitStruct;
+
+#[test]
+fn unit_struct_into_value() {
+    let expected = Value::test_nothing();
+    let actual = UnitStruct.into_value_unknown();
+    assert_eq!(expected, actual);
+}
+
+#[test]
+fn unit_struct_from_value() {
+    let expected = UnitStruct;
+    let actual = UnitStruct::from_value(Value::test_nothing(), Span::test_data()).unwrap();
+    assert_eq!(expected, actual);
+}
+
+#[test]
+fn unit_struct_roundtrip() {
+    let expected = UnitStruct;
+    let actual =
+        UnitStruct::from_value(UnitStruct.into_value_unknown(), Span::test_data()).unwrap();
+    assert_eq!(expected, actual);
+
+    let expected = Value::test_nothing();
+    let actual = UnitStruct::from_value(Value::test_nothing(), Span::test_data())
+        .unwrap()
+        .into_value_unknown();
+    assert_eq!(expected, actual);
+}
+
+#[derive(IntoValue, FromValue, Debug, PartialEq)]
+enum Enum {
+    Unit,
+    Tuple(u32, String),
+    Struct { a: u32, b: String },
+}
+
+impl Enum {
+    fn make() -> [Self; 3] {
+        [
+            Enum::Unit,
+            Enum::Tuple(12, "Tuple variant".to_string()),
+            Enum::Struct {
+                a: 34,
+                b: "Struct variant".to_string(),
+            },
+        ]
+    }
+
+    fn value() -> Value {
+        Value::test_list(vec![
+            Value::test_record(record! {
+                "type" => Value::test_string("unit")
+            }),
+            Value::test_record(record! {
+                "type" => Value::test_string("tuple"),
+                "content" => Value::test_list(vec![
+                    Value::test_int(12),
+                    Value::test_string("Tuple variant")
+                ])
+            }),
+            Value::test_record(record! {
+                "type" => Value::test_string("struct"),
+                "content" => Value::test_record(record! {
+                    "a" => Value::test_int(34),
+                    "b" => Value::test_string("Struct variant")
+                })
+            }),
+        ])
+    }
```

Comment:
Q: should we provide a derive for `enum` if we don't have a canonical representation for that in the Nushell type system (yet) and we have to make one up here for completeness sake.

We kind of have enum-like patterns with options that only accept a set of strings, they could map to C-style enums. This at the same time would have a distinct encoding from a tagged-union as suggested here.

Reply:
We can, at least for now, drop the derive for enums and let users handle enums themselves, as they can also implement `FromValue` and `IntoValue` on their own. But it would be still very practical. Also `Option`s are kind of annoying since no `Value` variant depicts that a value is optional.

Reply:
As we don't make a non-null guarantee for outputs at the moment I think your `Option` implementation is good.
But yeah we make that distinction for command parameters (and field access via the optional/standard cell-path).